### DISCUSSION TITLE
list of Buttons

### DISCUSSION THREAD

**Discussion by JohanBenko** [gpiozero/gpiozero#955] [Q&A]
Hi all, I just started programming in python so basically I am not as smart as I look...
All examples are with one or two buttons but for now I ended up with 6 for my project.

```python
    from gpiozero import Button 

    buttons={"stop":4,"zoom":7,"menu":23,"plus":24,"minus":22,"exit":5}
    button_list=[]

    for button in buttons:
       button_list.append(Button(buttons[button]))
```

But that doesn't seem to work error: `self = super(GPIOMeta, cls).__call__(*args, **kwargs)`
What I am doing wrong?

python 3.7.3
gpiozero 1.5.1

**Comment by lurch** [comment]
I can't immediately see what you're doing wrong (and I'm not at a Pi right now), but as an unrelated suggestion perhaps you might want to e.g. change `button_list=[]` to `button_dict = {}` and then do `button_dict[button] = Button(buttons[button])` inside the loop?

**Comment by bennuttall** [comment]
Your code works fine for me:

```
In [3]: from gpiozero import Button
   ...: 
   ...: buttons={"stop":4,"zoom":7,"menu":23,"plus":24,"minus":22,"exit":5}
   ...: button_list=[]
   ...: 
   ...: for button in buttons:
   ...:    button_list.append(Button(buttons[button]))
   ...: 

In [4]: button_list
Out[4]: 
[<gpiozero.Button object on pin GPIO4, pull_up=True, is_active=False>,
 <gpiozero.Button object on pin GPIO7, pull_up=True, is_active=False>,
 <gpiozero.Button object on pin GPIO23, pull_up=True, is_active=False>,
 <gpiozero.Button object on pin GPIO24, pull_up=True, is_active=False>,
 <gpiozero.Button object on pin GPIO22, pull_up=True, is_active=False>,
 <gpiozero.Button object on pin GPIO5, pull_up=True, is_active=False>]
```

However, there are a few pieces of advice I want to offer.

Seeing as you don't seem to use the button names, you could just create a list of Buttons on those pins:

```python
buttons = [Button(pin) for pin in (4, 7, 23, 24, 22, 5)]
```

Or if you want to keep the dictionary:

```
button_names = {"stop": 4, "zoom": 7, "menu": 23, "plus": 24, "minus": 22, "exit": 5}

buttons = [Button(pin) for pin in button_names.values()]
```

It's worth knowing about different ways of iterating over dictionaries:

```python
d = {"A": 1, "B": 2}

for k in d:
    print(k)
# outputs A and B

for v in d.values():
    print(v)
# outputs 1 and 2

for k, v in d.items():
    print(k, v)
# outputs ('A', 1) and ('B', 2)
```

And finally, you could use `ButtonBoard`:

```python
bb = ButtonBoard(stop=4, zoom=7, menu=23, plus=24, minus=22, exit=5)
```

or you could even unpack your dictionary to the same effect:

```python
buttons = {"stop": 4, "zoom": 7, "menu": 23, "plus": 24, "minus":22, "exit": 5}

bb = ButtonBoard(**buttons)
```

Once you have a `ButtonBoard`, each button is available at e.g. `bb.stop`:

```python
bb.stop.when_pressed = stop
```

I'm not sure what's causing your error but perhaps try the simplest thing possible first, and work your way up and see what causes it to break. Start with opening a shell and running:

```
>>> from gpiozero import Button
>>> btn = Button(4)
```

Hope that helps!

**Reply by lurch** [reply]
Haha, good call with `ButtonBoard` :smiley:

**Reply by JohanBenko** [reply]
Thank  you @bennuttall , I still have to decrypt the ```ButtonBoard``` solution. Seems next level to me.
The shell commands run fine, no errors

The context of my code:
```
def button_pressed(button_name):
    button_index=list(buttons).index(button_name)
    if button_list[button_index].is_pressed:
        while button_list[button_index].is_pressed:
            sleep(0.05)
        return True
    else:
        return False 
```
that way I can theoretically check if a button is pressed with the following call (f.e. the 'menu' button:
```
    if button_pressed("menu"):
       ....
```
The problem seems to be linked to the instruction where I append the button to the list. If I add them line by line without the loop everything is fine.

**Reply by JohanBenko** [reply]
Update: I have built a new image. Exact same setup and code. Now python complains that button_list is not defined but the loop creating the button is accepted this time (why???).

**Reply by lurch** [reply]
Don't take this the wrong way, but that's a fairly horrid code-solution!

You don't want to "check if button1 is pressed, then check if button2 is pressed, then check if button3 is pressed, ..." you want to assign functions to each button's `when_pressed` or `when_released` event-handler. Have another read through of https://gpiozero.readthedocs.io/en/stable/recipes.html

**Reply by JohanBenko** [reply]
No problem I can handle some critics! At first I wanted to follow an approach similar to yours but I realised that not all buttons make sense at a given moment; I only check the buttons that are appropriate in the context. fe Exit has only a meaning when you are in the menus same as do minus and plus. Stop has only a meaning when you are not navigating in the menu etc... it is not like interfacing always with the same led or play a song assigned to a button. 
As I said before I am very noob and the examples are very basic... 
I also changed my mind after reading about debouncing the buttons (yes mine bounce). If I make an event-handler I have to check if it is not called by accident (due to bouncing) and whether the context is ok?

**Reply by lurch** [reply]
> yes mine bounce

*All* buttons bounce, yours aren't special :wink:  But there's an option in the Button constructor to turn on software debouncing - something like `bounce_time=0.01` should be sufficient?  @waveform80 & @bennuttall Perhaps one of the examples in the basic recipes page should show (and explain the need for) button debouncing, as it seems like something that might trip up beginners fairly often?

>  I only check the buttons that are appropriate in the context. fe Exit has only a meaning when you are in the menus

Probably the easiest way of solving that is to have a global variable, and then you could do something like:
```
in_menu = False

def enter_menu():
    global in_menu
    if not in_menu:
        in_menu = True

def exit_menu():
    global in_menu
    if in_menu:
        in_menu = False

def go_up():
    if in_menu:
        # do your "up" action here

def go_down():
    if in_menu:
        # do your "down" action here

menu_button.when_pressed = enter_menu
exit_button.when_pressed = exit_menu
plus_button.when_pressed = go_up
minus_button.when_pressed = go_down
```
etc. etc.
The `global in_menu` line is only needed in the functions in which you'll be changing the value of `in_menu`.

**Reply by JohanBenko** [reply]
@lurch thanks for gently pushing me in the right direction. Booking some progress now and will post the working solution for future reference.

**Reply by JohanBenko** [reply]
Thanks for all the help guys! I came up with the following working program. It is no coincidence that you might recognize some words or structure 😉
```
from gpiozero import Button, ButtonBoard
from picamera import PiCamera, Color
from os import system
from time import sleep

camera=PiCamera()
camera.resolution =(1440,1080)
camera.start_preview()
camera.annotate_background = Color('black')
camera.annotate_foreground = Color('white')
camera.annotate_text_size = 50

effects =["none","negative","solarize",
          "sketch","denoise","oilpaint",
          "hatch","gpen","pastel",
          "watercolor","film","blur",
          "saturation","colorswap",
          "washedout","posterise","colorpoint",
          "colorbalance","cartoon",
          "deinterlace1","deinterlace2"]
exposuremodes=["off","auto","night","nightpreview",
          "backlight","spotlight","sports",
          "snow","beach","verylong","fixedfps",
          "antishake","fireworks"]
wbmodes=["off","auto","sunlight","cloudy",
         "shade","tungsten","fluorescent",
         "incandescent","flash","horizon"]
menu_items=["Effect","Exposure","WB","Brightness","Contrast","Zoom"]

menu_item_len=[len(effects)-1,len(exposuremodes)-1,len(wbmodes)-1,99,99,99]
data_values=[3,1,1,50,70,40]
menu_len = len(menu_items)-1
menu_item = 0
in_menu = True

buttons = {"stop": 4, "menu": 23, "up": 24, "down":22, "exit": 5, "zoom":17}
button = ButtonBoard(**buttons)

def show_menu():
    if menu_item == 0:        
        camera.image_effect = effects[data_values[menu_item]]
        camera.annotate_text = menu_items[menu_item]+": "+effects[data_values[menu_item]] 
    elif  menu_item == 1:        
        camera.exposure_mode = exposuremodes[data_values[menu_item]]
        camera.annotate_text = menu_items[menu_item]+": "+exposuremodes[data_values[menu_item]]
    elif menu_item == 2:
        camera.awb_mode = wbmodes[data_values[menu_item]]
        camera.annotate_text = menu_items[menu_item]+": "+wbmodes[data_values[menu_item]] 
    elif menu_item == 3:
        camera.brightness=data_values[menu_item]
        camera.annotate_text = menu_items[menu_item]+": "+str(data_values[menu_item])
    elif menu_item == 4:
        camera.contrast=data_values[menu_item] 
        camera.annotate_text = menu_items[menu_item]+": "+str(data_values[menu_item]) 
    elif menu_item == 5:
        camera.zoom = (data_values[menu_item]/200,data_values[menu_item]/200,data_values[menu_item]/100,data_values[menu_item]/100)
        camera.annotate_text = menu_items[menu_item]+": "+str(data_values[menu_item])

def enter_menu():
    global in_menu
    global menu_item
    if not in_menu:
        in_menu = True
    else:
        menu_item+=1
        if menu_item>menu_len:
            menu_item = 0
    show_menu()
            
def exit_menu():
    global in_menu
    if in_menu:
        in_menu = False
        camera.annotate_text=""
        camera.zoom = (.0,.0,1.,1.)

def up():
    global data_values
    if in_menu and data_values[menu_item] < menu_item_len[menu_item]:
        data_values[menu_item]+=1
        show_menu()
        
def down():
    global data_values
    if in_menu and data_values[menu_item] > 0:
        data_values[menu_item]-=1
        show_menu()
        
def stop_system():
    camera.stop_preview()
    camera.close()           
    #system("shutdown now -h") #uncomment to shutdown the Pi
    
def zoom():
    zoom_factor=data_values[5]/100
    camera.zoom = (zoom_factor/2,zoom_factor/2,zoom_factor,zoom_factor)
    while button.zoom.is_pressed:
        sleep(0.05)
    camera.zoom = (.0,.0,1.,1.)

for menu_item in range(menu_len):
    show_menu()
exit_menu()
menu_item = 0

button.menu.when_pressed = enter_menu
button.exit.when_pressed = exit_menu
button.up.when_pressed = up
button.down.when_pressed = down
button.stop.when_pressed = stop_system
button.zoom.when_pressed = zoom

while True:
    sleep(0.05)

```
Notice I kept the 'button names' in the dict for future use (f.e. for logging the key presses)

**Reply by lurch** [reply]
Cool, glad to hear you managed to get it all working :+1: 

BTW rather than doing `menu_len = len(menu_items)-1` and then `if menu_item>menu_len:`, it _might_ be cleaner to do `menu_len = len(menu_items)` and then `if menu_item >= menu_len:` :slightly_smiling_face:

**Reply by bennuttall** [reply]
It's always better to iterate over the items rather than iterate over numbers in the range of indexes, e.g.

```
for thing in things:
    do(thing)
```

**Reply by JohanBenko** [reply]
> 
> 
> Cool, glad to hear you managed to get it all working 👍
> 
> BTW rather than doing `menu_len = len(menu_items)-1` and then `if menu_item>menu_len:`, it _might_ be cleaner to do `menu_len = len(menu_items)` and then `if menu_item >= menu_len:` 🙂

When I was young (...) I have been taught never to use '=' ... that's the reason. 
If I recall well it was for two reasons:
- execution time
- the obvious reason when working with floats (not the case here)

In that time it was called 'good practice'.

**Reply by JohanBenko** [reply]
> 
> 
> It's always better to iterate over the items rather than iterate over numbers in the range of indexes, e.g.
> 
> ```
> for thing in things:
>     do(thing)
> ```

Seems nice and lean but where could this be applied?
Do you mean for
```
for menu_item in range(menu_len):
    show_menu()
```
?

**Reply by JohanBenko** [reply]
For completeness and future reference: *Sometimes* (why?) the zoom doesn't work. This could be solved by replacing
```
zoom_factor=data_values[5]/100
```
by
```
zoom_factor=float(data_values[5]/100)
```

**Reply by lurch** [reply]
In Python2, the `/` operator does integer-division if both arguments are integers, so `50 / 100 == 0` and `50.0 / 100 == 0.5`. In Python3 it always does floating-point-division so `50 / 100 == 0.5` and `50.0 / 100 == 0.5`.

Another top tip: Rather than have your `data_values` as a list, and accessing items by index (e.g. `data_values[5]`), it would be easier to make it a dict and then access items by name (e.g. `data_values['Zoom']`). :slightly_smiling_face:
